// @base-ui-components/react/dialog/store.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DialogStore = void 0;
var React = _interopRequireWildcard(require("react"));
var _store = require("@base-ui-components/utils/store");
const selectors = {
  open: (0, _store.createSelector)(state => state.open),
  modal: (0, _store.createSelector)(state => state.modal),
  nested: (0, _store.createSelector)(state => state.nested),
  nestedOpenDialogCount: (0, _store.createSelector)(state => state.nestedOpenDialogCount),
  dismissible: (0, _store.createSelector)(state => state.dismissible),
  openMethod: (0, _store.createSelector)(state => state.openMethod),
  descriptionElementId: (0, _store.createSelector)(state => state.descriptionElementId),
  titleElementId: (0, _store.createSelector)(state => state.titleElementId),
  mounted: (0, _store.createSelector)(state => state.mounted),
  transitionStatus: (0, _store.createSelector)(state => state.transitionStatus),
  triggerProps: (0, _store.createSelector)(state => state.triggerProps),
  popupProps: (0, _store.createSelector)(state => state.popupProps),
  floatingRootContext: (0, _store.createSelector)(state => state.floatingRootContext),
  popupElement: (0, _store.createSelector)(state => state.popupElement),
  triggerElement: (0, _store.createSelector)(state => state.triggerElement)
};
class DialogStore extends _store.ReactStore {
  static create(initialState) {
    const context = {
      popupRef: /*#__PURE__*/React.createRef(),
      backdropRef: /*#__PURE__*/React.createRef(),
      internalBackdropRef: /*#__PURE__*/React.createRef()
    };
    return new DialogStore(initialState, context, selectors);
  }
  setOpen = (nextOpen, eventDetails) => {
    this.context.openChange?.(nextOpen, eventDetails);
    if (eventDetails.isCanceled) {
      return;
    }
    const details = {
      open: nextOpen,
      nativeEvent: eventDetails.event,
      reason: eventDetails.reason,
      nested: this.state.nested
    };
    this.state.floatingRootContext.events?.emit('openchange', details);
    this.set('open', nextOpen);
  };
}
exports.DialogStore = DialogStore;
