// @base-ui-components/react/alert-dialog/root/AlertDialogRoot.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AlertDialogRoot = void 0;
var React = _interopRequireWildcard(require("react"));
var _useRefWithInit = require("@base-ui-components/utils/useRefWithInit");
var _useDialogRoot = require("../../dialog/root/useDialogRoot");
var _store = require("../../dialog/store");
var _DialogRootContext = require("../../dialog/root/DialogRootContext");
var _useFloatingRootContext = require("../../floating-ui-react/hooks/useFloatingRootContext");
var _constants = require("../../utils/constants");
var _jsxRuntime = require("react/jsx-runtime");
const INITIAL_STATE = {
  open: false,
  dismissible: false,
  nested: false,
  popupElement: null,
  triggerElement: null,
  modal: true,
  descriptionElementId: undefined,
  titleElementId: undefined,
  openMethod: null,
  mounted: false,
  transitionStatus: 'idle',
  nestedOpenDialogCount: 0,
  triggerProps: _constants.EMPTY_OBJECT,
  popupProps: _constants.EMPTY_OBJECT,
  floatingRootContext: (0, _useFloatingRootContext.getEmptyContext)()
};

/**
 * Groups all parts of the alert dialog.
 * Doesn’t render its own HTML element.
 *
 * Documentation: [Base UI Alert Dialog](https://base-ui.com/react/components/alert-dialog)
 */
const AlertDialogRoot = exports.AlertDialogRoot = function AlertDialogRoot(props) {
  const {
    children,
    defaultOpen = false,
    onOpenChange,
    onOpenChangeComplete,
    open: openProp,
    actionsRef
  } = props;
  const parentDialogRootContext = (0, _DialogRootContext.useDialogRootContext)();
  const nested = Boolean(parentDialogRootContext);
  const store = (0, _useRefWithInit.useRefWithInit)(_store.DialogStore.create, INITIAL_STATE).current;
  store.useControlledProp('open', openProp, defaultOpen);
  store.useSyncedValue('nested', nested);
  store.useContextCallback('openChange', onOpenChange);
  store.useContextCallback('openChangeComplete', onOpenChangeComplete);
  (0, _useDialogRoot.useDialogRoot)({
    store,
    actionsRef,
    parentContext: parentDialogRootContext?.store.context,
    onOpenChange
  });
  const contextValue = React.useMemo(() => ({
    store
  }), [store]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_DialogRootContext.DialogRootContext.Provider, {
    value: contextValue,
    children: children
  });
};
if (process.env.NODE_ENV !== "production") AlertDialogRoot.displayName = "AlertDialogRoot";
