// @base-ui-components/react/scroll-area/root/ScrollAreaRootCssVars.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScrollAreaRootCssVars = void 0;
let ScrollAreaRootCssVars = exports.ScrollAreaRootCssVars = /*#__PURE__*/function (ScrollAreaRootCssVars) {
  /**
   * The scroll area's corner height.
   * @type {number}
   */
  ScrollAreaRootCssVars["scrollAreaCornerHeight"] = "--scroll-area-corner-height";
  /**
   * The scroll area's corner width.
   * @type {number}
   */
  ScrollAreaRootCssVars["scrollAreaCornerWidth"] = "--scroll-area-corner-width";
  /**
   * The distance from the horizontal start edge in pixels.
   * @type {number}
   */
  ScrollAreaRootCssVars["scrollAreaOverflowXStart"] = "--scroll-area-overflow-x-start";
  /**
   * The distance from the horizontal end edge in pixels.
   * @type {number}
   */
  ScrollAreaRootCssVars["scrollAreaOverflowXEnd"] = "--scroll-area-overflow-x-end";
  /**
   * The distance from the vertical start edge in pixels.
   * @type {number}
   */
  ScrollAreaRootCssVars["scrollAreaOverflowYStart"] = "--scroll-area-overflow-y-start";
  /**
   * The distance from the vertical end edge in pixels.
   * @type {number}
   */
  ScrollAreaRootCssVars["scrollAreaOverflowYEnd"] = "--scroll-area-overflow-y-end";
  return ScrollAreaRootCssVars;
}({});
